### PR TITLE
ARROW-16749: [Go] Fix pqarrow writer for null array

### PR DESCRIPTION
**Pull request by alexandreyc**
Hello world,

When converting from Arrow to Parquet it looks like there is a bug with arrays of type `arrow.NULL`. Here is a snippet of code to reproduce the bug:
```go
package main

import (
	"fmt"
	"log"
	"os"

	"github.com/apache/arrow/go/v9/arrow"
	"github.com/apache/arrow/go/v9/arrow/array"
	"github.com/apache/arrow/go/v9/arrow/memory"
	"github.com/apache/arrow/go/v9/parquet/pqarrow"
)

const n = 10

func run() error {
	schema := arrow.NewSchema(
		[]arrow.Field{
			{Name: "f1", Type: arrow.Null, Nullable: true},
		},
		nil,
	)

	rb := array.NewRecordBuilder(memory.DefaultAllocator, schema)
	defer rb.Release()

	for i := 0; i < n; i++ {
		rb.Field(0).(*array.NullBuilder).AppendNull()
	}

	rec := rb.NewRecord()
	defer rec.Release()

	for i, col := range rec.Columns() {
		fmt.Printf("column[%d] %q: %v\n", i, rec.ColumnName(i), col)
	}

	f, err := os.Create("output.parquet")
	if err != nil {
		return err
	}
	defer f.Close()

	w, err := pqarrow.NewFileWriter(rec.Schema(), f, nil, pqarrow.DefaultWriterProps())
	if err != nil {
		return err
	}
	defer w.Close()

	err = w.Write(rec)
	if err != nil {
		return err
	}

	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}
```

I'm not 100% sure if my fix is the best, so feel free to correct me.

Thanks

A

### CODE REVIEW

**Comment by alexandreyc** [comment]
Yes, it doesn't look like there is an equivalent Parquet type. Nonetheless, it seems that decision to convert Arrow's `Null` to Parquet's `Int32` has already been made in the code base. See [here](https://github.com/apache/arrow/blob/8c63788ff7d52812599a546989b7df10887cb01e/go/parquet/pqarrow/schema.go#L259). 

The use case is quite simple: in real world databases there are columns which are full of `NULL` values and I guess that's why Arrow has `Null`. Some might say that's a bad practice but it exists...

I don't see any reason why we should not allow converting record batches with `Null` arrays to Parquet.

What do you think?

In the mean time I will look into others implementations to see what's being done here.

**Comment by zeroshade** [comment]
@alexandreyc Huh, I didn't even remember that I did that. Looks like parquet does have a `NULL` logical type so it makes sense.

That said, instead of allocating an entire int32 slice for this. I believe the better solution would be to manually call `wr.WriteBatchSpaced(nil, defLevels, repLevels, learArr.NullBitmapBytes(), 0)` and then break. No need to allocate an entire int32 slice for this case.

**Comment by alexandreyc** [comment]
@zeroshade thanks for your answer.

- I added a test to exercise the bug
- I had to fix an issue with column writers (feel free to double check since I'm not very familiar with the parquet's code base)
- I regenerate parquet's code (it looks like there has been modifications to metadata statistics that hasn't been taken into account)

File: go/parquet/file/column_writer_types.gen.go.tmpl [hunk 1]
```diff
@@ -137,13 +137,13 @@ func (w *{{.Name}}ColumnChunkWriter) WriteBatchSpaced(values []{{.name}}, defLev
 
     w.writeLevelsSpaced(batch, levelSliceOrNil(defLevels, offset, batch), levelSliceOrNil(repLevels, offset, batch))
     if values != nil {
-      vals = values[valueOffset:]
+      vals = values[valueOffset:][:info.numSpaced()]
```
Comment (zeroshade) [inline go/parquet/file/column_writer_types.gen.go.tmpl:140]: why was this needed? also, if you're going to add this it should be `vals = values[valueOffset : info.numSpaced()]` no need for the extra set of brackets.

**Review by zeroshade — CHANGES_REQUESTED** [review]

File: go/parquet/file/column_writer_types.gen.go.tmpl [hunk 2]
```diff
@@ -137,13 +137,13 @@ func (w *{{.Name}}ColumnChunkWriter) WriteBatchSpaced(values []{{.name}}, defLev
...
     }
 
     if w.bitsBuffer != nil {
-      w.writeValuesSpaced(vals[:info.numSpaced()], info.batchNum, w.bitsBuffer.Bytes(), 0)
+      w.writeValuesSpaced(vals, info.batchNum, w.bitsBuffer.Bytes(), 0)
     } else {
-      w.writeValuesSpaced(vals[:info.numSpaced()], info.batchNum, validBits, validBitsOffset+valueOffset)
+      w.writeValuesSpaced(vals, info.batchNum, validBits, validBitsOffset+valueOffset)
```
[the remark this answers is not in the document]
Comment (alexandreyc) [reply go/parquet/file/column_writer_types.gen.go.tmpl:146]: Yes that's the reason. I'm not sure it will change anything to replace the nil slice by an empty one because we are slicing out of bounds anyway.  `vals[:info.numSpaced()]` will panic in both cases (nil and empty) because in that case `info.numSpaced() > 0` (because null count > 0). 

(Or maybe I misunderstood your message)

File: go/parquet/file/column_writer_types.gen.go.tmpl (hunk 1 above)
Comment (alexandreyc) [reply go/parquet/file/column_writer_types.gen.go.tmpl]: I made the change to use the following form: `vals = values[valueOffset:valueOffset+info.numSpaced()]`

File: go/parquet/file/column_writer_types.gen.go.tmpl (hunk 2 above)
[the remark this answers is not in the document]
Comment (zeroshade) [reply go/parquet/file/column_writer_types.gen.go.tmpl:146]: Gotcha, that would be my mistake actually i forgot that `numSpaced` would be the full number, not zero which would be fine.

**Review by zeroshade — APPROVED** [review]

**Comment by zeroshade** [comment]
@alexandreyc Looks good to me now! Thanks much for this!
